Add `FieldsN` function to `strings` package.

celestix opened the pull request:
FieldsN splits a string into n fields.
It is similar to strings.Fields, but it splits the string into n fields.
If the string has less than n fields, the last field will contain the rest of the string.
If the string has more than n fields, the last field will contain the rest of the string.
If n is -1, it will behave like strings.Fields.

raklaptudirm commented on strings/fieldsn.go:
```diff
@@ -0,0 +1,37 @@
+			b.WriteRune(c)
+			continue
+		}
+		switch c {
```
Instead of a switch, `unicode.IsSpace(r)` should be better.

raklaptudirm commented on strings/fieldsn.go:
```diff
@@ -0,0 +1,37 @@
+		return strings.Fields(s)
+	}
+	a := make([]string, n+1)
+	b := strings.Builder{}
```
A string builder seems overkill in this situation, just a simple string should suffice.

raklaptudirm commented on strings/fieldsn.go:
```diff
@@ -0,0 +1,37 @@
+// If n is -1, it will behave like strings.Fields.
+func FieldsN(s string, n int) []string {
+	if n < 0 {
+		return strings.Fields(s)
```
Change the algorithm so that you don't depend on the standard library implementation.

souhardya79 commented:
I want to contribute in this issue

celestix commented:
@raklaptudirm I've made the changes you asked, please check if it is fine now.

raklaptudirm commented on strings/fieldsn.go:
```diff
@@ -0,0 +1,38 @@
+	}
+	a := make([]string, n+1)
+	b := ""
+	index := 0
```
Instead of an index variable, you can just use the first value from the range expression.

celestix replied:
that index variable is not the index for character element in the string, It's incremented on every whitespace.

raklaptudirm commented on strings/fieldsn.go:
```diff
@@ -0,0 +1,38 @@
+	b := ""
+	index := 0
+	for _, c := range s {
+		if index+1 > n {
```
This is equivalent to `index >= n`

raklaptudirm commented on strings/fieldsn.go:
```diff
@@ -0,0 +1,38 @@
+			b += string(c)
+			continue
+		}
+		if !unicode.IsSpace(c) {
```
Both the if statements have the same body, so you should merge the conditions with a logical or `||`.

celestix replied:
Sure!

celestix commented:
Hey @raklaptudirm, please check now.

raklaptudirm reviewed (changes requested):
The code doesn't properly handle cases where there are multiple spaces between fields and cases where there are leading and trailing spaces.

celestix commented:
@raklaptudirm Thanks for pointing out the multiple whitespaces issue! I've fixed it and added few testcases for those conditions too, please check if everything is fine now or not.

raklaptudirm reviewed (changes requested):
Instead of the wf variable, you could just check if `b == ""`.

Also note the if `NumFields(s) <= n` then also `len(n) == n + 1`, which is probably unintended.
